string.c

```diff
@@ -3303,12 +3303,27 @@ rb_str_buf_cat_ascii(VALUE str, const char *ptr)
     }
 }
 
+static inline bool
+str_enc_asciicompat_from_index(int index)
```

Comment:
I'm tempted to have a fast function / macro that just tell us if the string is `ASCII | UTF_8 | US_ASCII`.

It would be useful to go into fast paths, and I expect 99.99% of strings to be one of these 3 encodings. I think it could be implemented as a simple bitmask check on the RValue flags.

Reply:
> I'm tempted to have a fast function / macro that just tell us if the string is `ASCII | UTF_8 | US_ASCII`.

What is the difference between ASCII and US_ASCII?

Maybe you mean ASCII-compatible rather than ASCII?

Reply:
`ASCII` is `Encoding::ASCII_8BIT` aka `Encoding::BINARY`, so it more or less means: "no encoding but can be coerced to ASCII compatible if the content fits".

`US_ASCII` is `Encoding::US_ASCII` so true, 7bit ASCII.

Reply:
> `ASCII` is `Encoding::ASCII_8BIT` aka `Encoding::BINARY`, so it more or less means: "no encoding but can be coerced to ASCII compatible if the content fits".

Please *never* call ASCII_8BIT just "ASCII". It may lead to dangerous misunderstandings.
Also, please note that there are many other "ASCII-compatible" encodings, and all of them can be simplified to "US_ASCII" if they only contain data with the highest bit in each byte set to 0 (which I'm guessing you mean by "the content fits").

> `US_ASCII` is `Encoding::US_ASCII` so true, 7bit ASCII.

That one is fine.

Reply:
Well, that's not me. That constant has existed for years. I'm more than in favor of renaming all that I even submitted a feature request and a PR, but it wasn't accepted.

Reply:
here you go: https://github.com/ruby/ruby/pull/6127

Reply:
> please note that there are many other "ASCII-compatible" encodings, and all of them can be simplified to "US_ASCII" if they only contain data with the highest bit in each byte set to 0 (which I'm guessing you mean by "the content fits").

Yes I know. What I'm suggesting is to have a fast-path for these 3 very common encodings, not to change any semantic. My profiling shows that constantly having to check the `rb_encoding` struct waste a lot of perf. If we could have a fastpath that just check the `encindex` we could optimize string operations quite a lot.